internal/expr/expr_test.go

```diff
@@ -326,44 +413,136 @@ func FuzzParser(f *testing.F) {
 	})
 }
 
-func (s *ExprSuite) TestValidPrepare(c *C) {
+func (s *ExprSuite) TestPrepareMismatchedStructName(c *C) {
 	testList := []struct {
-		input            string
-		prepareArgs      []any
-		expectedPrepared string
+		sql     string
+		structs []any
 	}{{
-		"SELECT street FROM t WHERE x = $Address.street",
-		[]any{Address{}},
-		"SELECT street FROM t WHERE x = ?",
+		sql:     "SELECT street FROM t WHERE x = $Address.street",
+		structs: []any{Person{ID: 1}},
 	}, {
-		"SELECT p FROM t WHERE x = $Person.id",
-		[]any{Person{}},
-		"SELECT p FROM t WHERE x = ?",
+		sql:     "SELECT street AS &Address.street FROM t",
+		structs: []any{},
+	}, {
+		sql:     "SELECT street AS &Address.id FROM t",
+		structs: []any{Person{ID: 1}},
 	}}
-	for _, test := range testList {
+
+	for i, test := range testList {
+		parser := expr.NewParser()
+		parsedExpr, err := parser.Parse(test.sql)
+		if err != nil {
+			c.Fatal(err)
+		}
+		_, err = parsedExpr.Prepare(test.structs...)
+		c.Assert(err, ErrorMatches, `cannot prepare expression: type [a-zA-Z_\.0-9]+ unknown, have: .*`,
+			Commentf("test %d failed:\nsql: '%s'\nstructs: '%+v'", i, test.sql, test.structs))
```

Comment:
It would be nice to have the real error here, as that's the only way we'll know the error message is indeed correct. As it is, even if it's right today, we can break it tomorrow and not notice.

For that, have you considered adding an `error string` field on the big testing list above, and if it's set, we fail unless an error is found, and do `ErrroMatches` with the set string?

Reply:
Changed.